addons/sale_timesheet/models/account_move.py

```diff
@@ -63,26 +63,45 @@ def action_view_timesheet(self):
             }
         }
 
-    def _link_timesheets_to_invoice(self, start_date=None, end_date=None):
+    def _link_timesheets_to_invoice(self, start_date=None, end_date=None, service_policies=('delivered_timesheet')):
         """ Search timesheets from given period and link this timesheets to the invoice
 
             When we create an invoice from a sale order, we need to
             link the timesheets in this sale order to the invoice.
             Then, we can know which timesheets are invoiced in the sale order.
             :param start_date: the start date of the period
             :param end_date: the end date of the period
+            :param service_policies: the service policies of the sol products to consider
         """
         for line in self.filtered(lambda i: i.move_type == 'out_invoice' and i.state == 'draft').invoice_line_ids:
-            sale_line_delivery = line.sale_line_ids.filtered(lambda sol: sol.product_id.invoice_policy == 'delivery' and sol.product_id.service_type == 'timesheet')
+            sale_line_delivery = line.sale_line_ids.filtered(lambda sol: sol.product_id.service_policy in service_policies if sol.product_id.service_policy else False)
             if sale_line_delivery:
                 domain = line._timesheet_domain_get_invoiced_lines(sale_line_delivery)
                 if start_date:
                     domain = expression.AND([domain, [('date', '>=', start_date)]])
+                if sale_line_delivery.product_id.service_policy in ('ordered_prepaid', 'delivered_manual'):
+                    end_date = line.move_id.invoice_date if line.move_id.invoice_date else fields.Date.today()
                 if end_date:
                     domain = expression.AND([domain, [('date', '<=', end_date)]])
                 timesheets = self.env['account.analytic.line'].sudo().search(domain)
                 timesheets.write({'timesheet_invoice_id': line.move_id.id})
 
+    def _unlink_timesheets_to_invoice(self, service_policies):
+        for move in self.filtered(lambda i: i.move_type == 'out_invoice' and i.state == 'posted'):
```

Comment:
```suggestion
        for move in self:
            if move.move_type == 'out_invoice' and move.state == 'posted':
```
filter loops on all moves. so, rather than looping twice on the same record set, we can do the check in a unique loop